File: android/apps/AutoTesterAndroid/app/src/main/java/com/mousebirdconsulting/autotester/TestCases/TextureVectorTestCase.java
package com.mousebirdconsulting.autotester.TestCases;

import android.app.Activity;
import android.content.res.AssetManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;

import com.mousebird.maply.ComponentObject;
import com.mousebird.maply.GlobeController;
import com.mousebird.maply.MapController;
import com.mousebird.maply.MaplyBaseController;
import com.mousebird.maply.MaplyTexture;
import com.mousebird.maply.Point2d;
import com.mousebird.maply.VectorInfo;
import com.mousebird.maply.VectorObject;
import com.mousebirdconsulting.autotester.Framework.MaplyTestCase;
import com.mousebirdconsulting.autotester.R;

import org.apache.commons.io.IOUtils;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.Charset;
import java.util.ArrayList;

public class TextureVectorTestCase extends MaplyTestCase
{
    private ArrayList<VectorObject> vectors = new ArrayList<VectorObject>();

    public TextureVectorTestCase(Activity activity) {
        super(activity);

        setTestName("Texture Vectors");
        setDelay(4);
        this.implementation = TestExecutionImplementation.Both;
    }

    static double ClipGridSize = 2.0/180.0*Math.PI;

    private void overlayCountries(MaplyBaseController baseVC,boolean globeMode) throws Exception {
        AssetManager assetMgr = getActivity().getAssets();
        String[] paths = assetMgr.list("country_json_50m");
        for (String path : paths) {
            InputStream stream = assetMgr.open("country_json_50m/" + path);
            try {
                VectorObject vecObject = new VectorObject();
                String json = IOUtils.toString(stream, Charset.defaultCharset());
                if (vecObject.fromGeoJSON(json)) {
                    vecObject.selectable = true;

                    // Work through each invidiual loop (hopefully)
                    for (VectorObject thisVecObj : vecObject)
                    {
                        VectorObject tessObj = null;
                        Point2d center = thisVecObj.centroid();

                        if (globeMode)
                        {
                            // We adjust the grid clipping size based on the latitude
                            // This helps a lot near the poles.  Otherwise we're way oversampling
                            float thisClipGridLon = (float)ClipGridSize;
                            if (Math.abs(center.getY()) > 60.0/180.0 * Math.PI)
                                thisClipGridLon *= 4.0;
                            else if (Math.abs(center.getY()) > 45.0/180.0 * Math.PI)
                                thisClipGridLon *= 2.0;

                            thisVecObj.getAttributes().setDouble("veccenterx",center.getX());
                            thisVecObj.getAttributes().setDouble("veccentery",center.getX());

                            // We clip the vector to a grid and then tesselate the results
                            // This forms the vector closer to the globe, make it look nicer
                            VectorObject clipped = thisVecObj.clipToGrid(new Point2d(thisClipGridLon, ClipGridSize));
                            if (clipped != null)
                                tessObj = clipped.tesselate();
                        }

                        if (tessObj == null)
                            tessObj = thisVecObj.tesselate();

                        vectors.add(tessObj);
                    }
                }
            } finally {
                try {
                    stream.close();
                } catch (IOException e) {
                }
            }
        }

        // Set up a dots texture
        Bitmap bm = BitmapFactory.decodeResource(activity.getResources(), R.drawable.dots);
        MaplyBaseController.TextureSettings texSettings = new MaplyBaseController.TextureSettings();
        texSettings.wrapU = true;  texSettings.wrapV = true;
        MaplyTexture tex = baseVC.addTexture(bm, texSettings, MaplyBaseController.ThreadMode.ThreadCurrent);

        // Add the vectors with textures
        VectorInfo vectorInfo = new VectorInfo();
        vectorInfo.setFilled(true);
        vectorInfo.setTexture(tex);
        vectorInfo.setTextureProjection(VectorInfo.TextureProjection.TangentPlane);
        vectorInfo.setTexScale(6.0,6.0);
        vectorInfo.setColor(Color.WHITE);
        ComponentObject compObj = baseVC.addVectors(vectors, vectorInfo, MaplyBaseController.ThreadMode.ThreadAny);
    }

    @Override
    public boolean setUpWithMap(MapController mapVC) throws Exception {
        StamenRemoteTestCase baseCase = new StamenRemoteTestCase(getActivity());
        baseCase.setUpWithMap(mapVC);
        overlayCountries(mapVC,false);
        return true;
    }

    @Override
    public boolean setUpWithGlobe(GlobeController globeVC) throws Exception {
        StamenRemoteTestCase baseCase = new StamenRemoteTestCase(getActivity());
        baseCase.setUpWithGlobe(globeVC);
        overlayCountries(globeVC,true);
        return true;
    }

    public ArrayList<VectorObject> getVectors() {
        return vectors;
    }}

File: android/apps/AutoTesterAndroid/app/src/main/java/com/mousebirdconsulting/autotester/TestCases/TextureVectorTestCase.kt
/*
 *  TextureVectorTestCase.java
 *  WhirlyGlobeLib
 *
 *  Created by Steve Gifford.
 *  Copyright 2011-2019 mousebird consulting
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
package com.mousebirdconsulting.autotester.TestCases

import android.app.Activity
import android.graphics.BitmapFactory
import android.graphics.Color
import android.util.Log
import com.mousebird.maply.*
import com.mousebirdconsulting.autotester.Framework.MaplyTestCase
import com.mousebirdconsulting.autotester.R
import okio.*
import java.lang.Exception
import java.nio.charset.Charset
import java.util.*
import kotlin.collections.ArrayList
import kotlin.math.PI
import kotlin.math.abs

class TextureVectorTestCase(activity: Activity) : MaplyTestCase(activity, "Textured Vectors") {
    
    // Grid size to use for clipping
    // Smaller is going to be more triangles, but look better
    // Might be better to use a bigger size for the poles
    private val ClipGridSize = 2.0/180.0* PI

    private fun buildCountries(control: BaseController) : Collection<ComponentObject> {
        val assetMgr = activity.assets
        val isGlobe = control is GlobeController
        
        val icons = arrayOf(
                BitmapFactory.decodeResource(activity.resources, R.drawable.dots),
                BitmapFactory.decodeResource(activity.resources, R.drawable.testtarget),
                BitmapFactory.decodeResource(activity.resources, R.drawable.sticker))
        val texs = icons.map { control.addTexture(it, null, ThreadMode.ThreadCurrent) }
        
        val paths = assetMgr.list("country_json_50m")?.sortedDescending() ?: emptyList()

        val tessObjs = arrayListOf(ArrayList<Pair<String,VectorObject>>())

        // Load in 20 countries to apply a texture to
        val numCountries = 20
        val numProjections = 3

        var count = 0
        for (path in paths) {
            try {
                val more = assetMgr.open("country_json_50m/$path").use { stream ->
                    val json = stream.source().use { source ->
                        source.buffer().use { buffer ->
                            buffer.readUtf8()
                        }
                    }
                    VectorObject.createFromGeoJSON(json)?.let { vecObj ->
                        vecObj.selectable = true
                        
                        // Work through each individual loop
                        for (thisVecObj in vecObj) {
                            val loopObj = thisVecObj.deepCopy()
            
                            // Center of the texture application
                            val center = loopObj.center()
                            val attrs = loopObj.attributes
                            attrs.setDouble("veccenterx", center.x)
                            attrs.setDouble("veccentery", center.y)

                            var thisClipGridLon = ClipGridSize
                            if (isGlobe) {
                                // We adjust the grid clipping size based on the latitude
                                // This helps a lot near the poles.  Otherwise we're way oversampling
                                if (abs(center.x) > 60.0 / 180.0 * PI)
                                    thisClipGridLon *= 4.0
                                else if (abs(center.y) > 45.0 / 180.0 * PI)
                                    thisClipGridLon *= 2.0;
                
                            }
                            loopObj.clipToGrid(Point2d(thisClipGridLon, ClipGridSize))?.tesselate()?.let {
                                val country = path.replace(".geojson","")
                                tessObjs.last().add(Pair(country,it))
                            }
                        }
                        count += 1
                    }
                    if (count >= numCountries) {
                        if (tessObjs.size >= numProjections) {
                            return@use false
                        }
                        tessObjs.add(ArrayList())
                        count = 0
                    }
                    return@use true
                }
                if (!more) {
                    break
                }
            } catch (e: Exception) {
                Log.e(javaClass.simpleName, "Failed to load $path", e)
            }
        }

        val projections = arrayOf(
            VectorInfo.TextureProjection.None,
            VectorInfo.TextureProjection.Screen,
            VectorInfo.TextureProjection.TangentPlane)

        return tessObjs.mapIndexed { index, items ->
            val vecInfo = VectorInfo().apply {
                setFilled(true)
                setTexture(texs[index])
                setTextureProjection(projections[index])
                setTexScale(16.0, 16.0)
                setColor(Color.WHITE)
            }
    
            val files = TreeSet(items.map { it.first }).joinToString(",")
            Log.i(javaClass.simpleName, "${projections[index]} => $files")
            
            // Add all the vectors at once to be more efficient
            // The geometry gets grouped together, which is nice and fast
            val objs = items.map { it.second }
            control.addVectors(objs, vecInfo, ThreadMode.ThreadAny)
        }
    }

    override fun setUpWithMap(mapVC: MapController): Boolean {
        baseCase.setUpWithMap(mapVC)
        vectorObjs.addAll(buildCountries(mapVC))
        mapVC.animatePositionGeo(-Math.PI/2,0.0,1.0,1.0)
        return true
    }

    override fun setUpWithGlobe(globeVC: GlobeController): Boolean {
        baseCase.setUpWithGlobe(globeVC)
        vectorObjs.addAll(buildCountries(globeVC))
        globeVC.animatePositionGeo(-Math.PI/2,0.0,1.0,1.0)
        return true
    }
    
    override fun shutdown() {
        controller?.removeObjects(vectorObjs, ThreadMode.ThreadCurrent)
        vectorObjs.clear()
        baseCase.shutdown()
        super.shutdown()
    }

    val baseCase = CartoLightTestCase(activity)
    val vectorObjs = ArrayList<ComponentObject>()
}